### ISSUE TITLE
Review #2

### ISSUE THREAD

**Issue by alt-romes** [liamzee/SPiDeW#1]
First of all, I want to say congratulations for what you have so far!

You modelled the program well, you made good choices on the monadic abstractions and even defined a type synonym `SPiDeW` for them, and defined program in terms of `SPiDeW`. I'm particularly happy to see this because you initially didn't understand Monad Transformers, and now you've used them quite well.

You've also defined a custom JSON instance where you needed the more fine grained control, and defined one using generics! The custom instance using applicative `<*>` and functors looks clean.

The IO http part seems to be going well, and you found out about `parseRequest` on your own, which is good (finding library functions without prior explanation)

Exploring your own style is also good, someone at ZuriHac did this project presentation that pointed out how one can be "creative" with the syntax layout :)

But I do have comments and notes on improvement:

https://github.com/liamzee/SPiDeW/blob/4e8dac3364c5d435544bebe12d34a064e8d6f55b/app/spidew.hs#L79-L80

This is a somewhat contrived function. When everything is put together, you have a special constructor `CommandNull` which you only use to convert to `Maybe` and then interpret, but it's confusing and it's basically going around in circles.

You approached the possibility of a "non-command" well, you modelled it with `CommandNull` as a data constructor of `Command`, but at the same time modelled the possibility of failure when parsing with `Maybe Command`. This isn't straightforward, and you can greatly simplify this story.

If you define a `CommandNull` constructor, then `parseInput` should just return `Command` (because any malformed command is a `CommandNull`, the `parseInput` function can be total if `CommandNull` is a constructor). Otherwise you can have `parseInput` return `Maybe Command`, but then you don't need `CommandNull`.

https://github.com/liamzee/SPiDeW/blob/4e8dac3364c5d435544bebe12d34a064e8d6f55b/app/spidew.hs#L52

This comment is another hint at something that might be better defined. It's good to *make invalid states unrepresentable*: therefore, in this case, either the constructor `CommandNull` is a valid state, or it shouldn't exist -- a comment in the data constructor about how this should only exist in some certain situations is a bit of a weak assertion.

Making this choice (you've used `Maybe` previously so I would suggest the other option in this assignment) will help you *get rid of ALL the weird things* that are happening regarding this:

* The `toMaybe` function is questionable
* the `toMaybe` is used
* The `executeCommand` has weird semantics: why does it receive maybe a command? you can get rid of the maybe altogether
* `parseInput` returns `Maybe Command`, but Command can already be `CommandNull`

The second thing I have to point, but is just a stylistic preference:

https://github.com/liamzee/SPiDeW/blob/4e8dac3364c5d435544bebe12d34a064e8d6f55b/app/spidew.hs#L182-L184

I would say having a type signature on top of the function really helps with readability: the type is very informative, and when reading your code (as a reader) I had to travel up and down a lot just to see what the type of the thing was.

To sum up,

Good work!

~romes

**Comment by liamzee** [comment]
Read, discussed privately, thanks for all the help, Romes!
